core/src/main/java/org/ehcache/config/builders/ConfigurationBuilder.java

```diff
@@ -0,0 +1,208 @@
+    if (existing == null) {
+      return withService(serviceConfiguration);
+    } else {
+      throw new IllegalArgumentException("There is already an instance of " + serviceConfiguration.getClass() + " registered: " + existing.getClass());
+    }
+  }
+
+  @Deprecated
+  public ConfigurationBuilder removeService(ServiceCreationConfiguration<?, ?> serviceConfiguration) {
+    List<ServiceCreationConfiguration<?, ?>> newServiceConfigurations = new ArrayList<>(serviceConfigurations);
+    newServiceConfigurations.remove(serviceConfiguration);
+    return new ConfigurationBuilder(this, newServiceConfigurations);
+  }
+
+  @Override
+  public <C extends ServiceCreationConfiguration<?, ?>> Collection<C> getServices(Class<C> configurationType) {
+    return serviceConfigurations.stream().filter(service -> configurationType.isAssignableFrom(service.getClass())).map(configurationType::cast).collect(toList());
+  }
+
+  @Override
+  public ConfigurationBuilder withService(ServiceCreationConfiguration<?, ?> config) {
+    List<ServiceCreationConfiguration<?, ?>> newServiceConfigurations = new ArrayList<>(serviceConfigurations);
+    newServiceConfigurations.removeIf(config.getClass()::isInstance);
+    newServiceConfigurations.add(config);
+    return new ConfigurationBuilder(this, newServiceConfigurations);
+  }
+
+  @Override
+  public <C extends ServiceCreationConfiguration<?, ?>> ConfigurationBuilder withoutServices(Class<C> clazz, Predicate<? super C> predicate) {
+    List<ServiceCreationConfiguration<?, ?>> newServiceConfigurations = new ArrayList<>(serviceConfigurations);
+    newServiceConfigurations.removeIf(c -> clazz.isInstance(c) && predicate.test(clazz.cast(c)));
+    return new ConfigurationBuilder(this, newServiceConfigurations);
+  }
+
+  @Override
+  public <R, C extends ServiceCreationConfiguration<?, R>> ConfigurationBuilder updateServices(Class<C> clazz, UnaryOperator<R> update) {
+    @SuppressWarnings("unchecked")
+    Collection<? extends ServiceCreationConfiguration<?, R>> existing = getServices(clazz);
+
+    if (existing.isEmpty()) {
+      throw new IllegalStateException("Cannot updates service configurations. No services exist");
+    } else {
+      ConfigurationBuilder otherBuilder = withoutServices(clazz);
+      for (ServiceCreationConfiguration<?, R> configuration : existing) {
+        otherBuilder = otherBuilder.withService(configuration.build(update.apply(configuration.derive())));
+      }
+      return otherBuilder;
+    }
+  }
+
+  @Override
+  public ClassLoader getClassLoader() {
+    return classLoader;
+  }
+
+  @Override
+  public ConfigurationBuilder withClassLoader(ClassLoader classLoader) {
+    return new ConfigurationBuilder(this, requireNonNull(classLoader));
+  }
+
+  @Override
+  public ConfigurationBuilder withoutClassLoader() {
+    return new ConfigurationBuilder(this, (ClassLoader) null);
+  }
+
+  @Deprecated
+  public boolean containsCache(String alias) {
```

Comment:
I presume that, even though this is a "new" class, it existed elsewhere before.  If there is existing non-internal code that uses this, it would benefit from having the `@deprecated` javadoc tag that indicates this method's replacement.

Reply:
Added javadoc.